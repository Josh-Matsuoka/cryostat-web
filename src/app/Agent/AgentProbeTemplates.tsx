/*
 * Copyright The Cryostat Authors
 *
 * The Universal Permissive License (UPL), Version 1.0
 *
 * Subject to the condition set forth below, permission is hereby granted to any
 * person obtaining a copy of this software, associated documentation and/or data
 * (collectively the "Software"), free of charge and under any and all copyright
 * rights in the Software, and any and all patent rights owned or freely
 * licensable by each licensor hereunder covering either (i) the unmodified
 * Software as contributed to or provided by such licensor, or (ii) the Larger
 * Works (as defined below), to deal in both
 *
 * (a) the Software, and
 * (b) any piece of software and/or hardware listed in the lrgrwrks.txt file if
 * one is included with the Software (each a "Larger Work" to which the Software
 * is contributed by such licensors),
 *
 * without restriction, including without limitation the rights to copy, create
 * derivative works of, display, perform, and distribute the Software and make,
 * use, sell, offer for sale, import, export, have made, and have sold the
 * Software and the Larger Work(s), and to sublicense the foregoing rights on
 * either these or other terms.
 *
 * This license is subject to the following condition:
 * The above copyright notice and either this complete permission notice or at
 * a minimum a reference to the UPL must be included in all copies or
 * substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
import * as React from 'react';
import { ServiceContext } from '@app/Shared/Services/Services';
import { NotificationCategory } from '@app/Shared/Services/NotificationChannel.service';
import { NO_TARGET } from '@app/Shared/Services/Target.service';
import { useSubscriptions } from '@app/utils/useSubscriptions';
import {
  ActionGroup,
  Button,
  FileUpload,
  Form,
  FormGroup,
  Modal,
  ModalVariant,
  Toolbar,
  ToolbarContent,
  ToolbarGroup,
  ToolbarItem,
  TextInput,
} from '@patternfly/react-core';
import { PlusIcon } from '@patternfly/react-icons';
import {
  Table,
  TableBody,
  TableHeader,
  TableVariant,
  IAction,
  IRowData,
  IExtraData,
  ISortBy,
  SortByDirection,
  sortable,
} from '@patternfly/react-table';
import { useHistory } from 'react-router-dom';
import { concatMap, filter, first } from 'rxjs/operators';
import { LoadingView } from '@app/LoadingView/LoadingView';
import { authFailMessage, ErrorView, isAuthFail } from '@app/ErrorView/ErrorView';
import { ProbeTemplate } from '@app/Shared/Services/Api.service';
import { DeleteWarningType } from '@app/Modal/DeleteWarningUtils';
import { DeleteWarningModal } from '@app/Modal/DeleteWarningModal';

export const AgentProbeTemplates = () => {
  const context = React.useContext(ServiceContext);
  const history = useHistory();

  const [templates, setTemplates] = React.useState([] as ProbeTemplate[]);
  const [filteredTemplates, setFilteredTemplates] = React.useState([] as ProbeTemplate[]);
  const [filterText, setFilterText] = React.useState('');
  const [modalOpen, setModalOpen] = React.useState(false);
  const [uploadFile, setUploadFile] = React.useState(undefined as File | undefined);
  const [uploadFilename, setUploadFilename] = React.useState('');
  const [uploading, setUploading] = React.useState(false);
  const [fileRejected, setFileRejected] = React.useState(false);
  const [sortBy, setSortBy] = React.useState({} as ISortBy);
  const [isLoading, setIsLoading] = React.useState(false);
  const [errorMessage, setErrorMessage] = React.useState('');
  const [rowDeleteData, setRowDeleteData] = React.useState({} as IRowData);
  const [warningModalOpen, setWarningModalOpen] = React.useState(false);
  const addSubscription = useSubscriptions();

  const tableColumns = React.useMemo(
    () => [
      { title: 'name', transforms: [sortable] },
      { title: 'xml', transforms: [sortable] },
    ],
    [sortable]
  );

  React.useEffect(() => {
    let filtered;
    if (!filterText) {
      filtered = templates;
    } else {
      const ft = filterText.trim().toLowerCase();
      filtered = templates.filter(
        (t: ProbeTemplate) => t.name.toLowerCase().includes(ft) || t.xml.toLowerCase().includes(ft)
      );
    }
    const { index, direction } = sortBy;
    if (typeof index === 'number') {
      const keys = ['name', 'xml'];
      const key = keys[index];
      const sorted = filtered.sort((a, b) => (a[key] < b[key] ? -1 : a[key] > b[key] ? 1 : 0));
      filtered = direction === SortByDirection.asc ? sorted : sorted.reverse();
    }
    setFilteredTemplates([...filtered]);
  }, [filterText, templates, sortBy]);

  const handleTemplates = React.useCallback(
    (templates) => {
      setTemplates(templates);
      setIsLoading(false);
      setErrorMessage('');
    },
    [setTemplates, setIsLoading, setErrorMessage]
  );

  const handleError = React.useCallback(
    (error) => {
      setIsLoading(false);
      setErrorMessage(error.message);
    },
    [setIsLoading, setErrorMessage]
  );

  const refreshTemplates = React.useCallback(() => {
    setIsLoading(true);
    addSubscription(
      context.target
        .target()
        .pipe(
          concatMap((target) => context.api.getProbeTemplates()),
          first()
        )
        .subscribe(
          (value) => handleTemplates(value),
          (err) => handleError(err)
        )
    );
  }, [addSubscription, context, context.target, context.api, setIsLoading, handleTemplates, handleError]);

  React.useEffect(() => {
    addSubscription(
      context.target.target().subscribe(() => {
        setFilterText('');
        refreshTemplates();
      })
    );
  }, [context, context.target, addSubscription, refreshTemplates]);

  React.useEffect(() => {
    if (!context.settings.autoRefreshEnabled()) {
      return;
    }
    const id = window.setInterval(
      () => refreshTemplates(),
      context.settings.autoRefreshPeriod() * context.settings.autoRefreshUnits()
    );
    return () => window.clearInterval(id);
  }, []);

  React.useEffect(() => {
    const sub = context.target.authFailure().subscribe(() => {
      setErrorMessage('Auth failure');
    });
    return () => sub.unsubscribe();
  }, [context.target]);

  const displayTemplates = React.useMemo(
    () => filteredTemplates.map((t: ProbeTemplate) => [t.name, t.xml]),
    [filteredTemplates]
  );

  const handleDelete = (rowData) => {
    addSubscription(
      context.api
        .deleteCustomProbeTemplate(rowData[0])
        .pipe(first())
        .subscribe(() => {
          refreshTemplates();
        })
    );
  };

  const handleInsert = (rowData) => {
    addSubscription(
      context.api
        .insertProbes(rowData[0])
        .pipe(first())
        .subscribe(() => {})
    );
  };

  const actionResolver = (rowData: IRowData, extraData: IExtraData) => {
    if (typeof extraData.rowIndex == 'undefined') {
      return [];
    }
    let actions = [
      {
        title: 'Insert Probes...',
        onClick: (event, rowId, rowData) => handleInsert(rowData),
      },
    ] as IAction[];
    actions = actions.concat([
      {
        isSeparator: true,
      },
      {
        title: 'Delete',
        onClick: (event, rowId, rowData) => handleDeleteButton(rowData),
      },
    ]);
    return actions;
  };

  const handleModalToggle = () => {
    setModalOpen((v) => {
      if (v) {
        setUploadFile(undefined);
        setUploadFilename('');
        setUploading(false);
      }
      return !v;
    });
  };

  const handleFileChange = (value, filename) => {
    setFileRejected(false);
    setUploadFile(value);
    setUploadFilename(filename);
  };

  const handleUploadSubmit = () => {
    if (!uploadFile) {
      window.console.error('Attempted to submit template upload without a file selected');
      return;
    }
    setUploading(true);
    addSubscription(
      context.api
        .addCustomProbeTemplate(uploadFile)
        .pipe(first())
        .subscribe((success) => {
          setUploading(false);
          if (success) {
            setUploadFile(undefined);
            setUploadFilename('');
            setModalOpen(false);
            refreshTemplates();
          }
        })
    );
  };

  React.useEffect(() => {
    addSubscription(
      context.notificationChannel
        .messages(NotificationCategory.ProbeTemplateUploaded)
        .subscribe((v) => refreshTemplates())
    );
  }, [addSubscription, context, context.notificationChannel, setTemplates]);

  React.useEffect(() => {
    addSubscription(
      context.notificationChannel.messages(NotificationCategory.TemplateDeleted).subscribe((v) => refreshTemplates())
    );
  }, [addSubscription, context, context.notificationChannel, setTemplates]);

  const handleUploadCancel = () => {
    setUploadFile(undefined);
    setUploadFilename('');
    setModalOpen(false);
  };

  const handleDeleteButton = React.useCallback(
    (rowData) => {
      if (context.settings.deletionDialogsEnabledFor(DeleteWarningType.DeleteEventTemplates)) {
        setRowDeleteData(rowData);
        setWarningModalOpen(true);
      } else {
        handleDelete(rowData);
      }
    },
    [context, context.settings, setWarningModalOpen, setRowDeleteData, handleDelete]
  );

  const handleWarningModalAccept = React.useCallback(() => {
    handleDelete(rowDeleteData);
  }, [handleDelete, rowDeleteData]);

  const handleWarningModalClose = React.useCallback(() => {
    setWarningModalOpen(false);
  }, [setWarningModalOpen]);

  const handleFileRejected = () => {
    setFileRejected(true);
  };

  const handleSort = (event, index, direction) => {
    setSortBy({ index, direction });
  };

  const authRetry = React.useCallback(() => {
    context.target.setAuthRetry();
  }, [context.target, context.target.setAuthRetry]);

  if (errorMessage != '') {
    return (
      <ErrorView
        title={'Error retrieving probe templates'}
        message={errorMessage}
        retry={isAuthFail(errorMessage) ? authRetry : undefined}
      />
    );
  } else if (isLoading) {
    return <LoadingView />;
  } else {
    return (
      <>
        <Toolbar id="event-templates-toolbar">
          <ToolbarContent>
            <ToolbarGroup variant="filter-group">
              <ToolbarItem>
                <TextInput
                  name="templateFilter"
                  id="templateFilter"
                  type="search"
                  placeholder="Filter..."
                  aria-label="Probe template filter"
                  onChange={setFilterText}
                />
              </ToolbarItem>
            </ToolbarGroup>
            <ToolbarGroup variant="icon-button-group">
              <ToolbarItem>
                <Button key="create" variant="primary" onClick={handleModalToggle}>
                  Upload
                </Button>
              </ToolbarItem>
            </ToolbarGroup>
            <DeleteWarningModal
              warningType={DeleteWarningType.DeleteProbeTemplates}
              visible={warningModalOpen}
              onAccept={handleWarningModalAccept}
              onClose={handleWarningModalClose}
            />
          </ToolbarContent>
        </Toolbar>
        <Table
          aria-label="Probe Templates Table"
          variant={TableVariant.compact}
          cells={tableColumns}
          rows={displayTemplates}
          actionResolver={actionResolver}
          sortBy={sortBy}
          onSort={handleSort}
        >
          <TableHeader />
          <TableBody />
        </Table>

        <Modal
          isOpen={modalOpen}
          variant={ModalVariant.large}
          showClose={true}
          onClose={handleModalToggle}
          title="Create Custom Probe Template"
          description="Create a customized probe template. This is a specialized XML file typically created using JDK Mission Control, which defines a set of events to inject and their options to configure."
        >
          <Form>
            <FormGroup
              label="Template XML"
              isRequired
              fieldId="template"
              validated={fileRejected ? 'error' : 'default'}
            >
              <FileUpload
                id="probetemplateName"
                value={uploadFile}
                filename={uploadFilename}
                onChange={handleFileChange}
                isLoading={uploading}
                validated={fileRejected ? 'error' : 'default'}
                dropzoneProps={{
                  accept: '.xml',
                  onDropRejected: handleFileRejected,
                }}
              />
            </FormGroup>
            <ActionGroup>
              <Button variant="primary" onClick={handleUploadSubmit} isDisabled={!uploadFilename}>
                Submit
              </Button>
              <Button variant="link" onClick={handleUploadCancel}>
                Cancel
              </Button>
            </ActionGroup>
          </Form>
        </Modal>
      </>
    );
  }
};
