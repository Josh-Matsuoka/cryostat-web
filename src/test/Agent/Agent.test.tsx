/*
 * Copyright The Cryostat Authors
 *
 * The Universal Permissive License (UPL), Version 1.0
 *
 * Subject to the condition set forth below, permission is hereby granted to any
 * person obtaining a copy of this software, associated documentation and/or data
 * (collectively the "Software"), free of charge and under any and all copyright
 * rights in the Software, and any and all patent rights owned or freely
 * licensable by each licensor hereunder covering either (i) the unmodified
 * Software as contributed to or provided by such licensor, or (ii) the Larger
 * Works (as defined below), to deal in both
 *
 * (a) the Software, and
 * (b) any piece of software and/or hardware listed in the lrgrwrks.txt file if
 * one is included with the Software (each a "Larger Work" to which the Software
 * is contributed by such licensors),
 *
 * without restriction, including without limitation the rights to copy, create
 * derivative works of, display, perform, and distribute the Software and make,
 * use, sell, offer for sale, import, export, have made, and have sold the
 * Software and the Larger Work(s), and to sublicense the foregoing rights on
 * either these or other terms.
 *
 * This license is subject to the following condition:
 * The above copyright notice and either this complete permission notice or at
 * a minimum a reference to the UPL must be included in all copies or
 * substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
import * as React from 'react';
import renderer, { act } from 'react-test-renderer';
import { render, screen, within } from '@testing-library/react';
import '@testing-library/jest-dom';
import { of } from 'rxjs';
import { EventTemplate, ProbeTemplate } from '@app/Shared/Services/Api.service';
import { MessageMeta, MessageType, NotificationMessage } from '@app/Shared/Services/NotificationChannel.service';
import { ServiceContext, defaultServices } from '@app/Shared/Services/Services';
import { EventTemplates } from '@app/Events/EventTemplates';
import userEvent from '@testing-library/user-event';
import { DeleteWarningType } from '@app/Modal/DeleteWarningUtils';
import { AgentProbeTemplates } from '@app/Agent/AgentProbeTemplates';

const mockConnectUrl = 'service:jmx:rmi://someUrl';
const mockTarget = { connectUrl: mockConnectUrl, alias: 'fooTarget' };

const mockMessageType = {type: "application", subtype: "json"} as MessageType;

const mockCustomEventTemplate: ProbeTemplate = {
    name: 'someProbeTemplate',
    xml: '<some><dummy><xml></xml></dummy></some>'
};

const mockAnotherTemplate = {...mockCustomEventTemplate, name: 'anotherProbeTemplate'}

const mockCreateTemplateNotification = { 
  meta: {
    category: 'ProbeTemplateUploaded',
    type: mockMessageType
  } as MessageMeta,
  message: { 
    template: mockAnotherTemplate 
  } 
} as NotificationMessage;
const mockDeleteTemplateNotification = 
{...mockCreateTemplateNotification, 
  meta: {
  category: 'ProbeTemplateDeleted',
  type: mockMessageType
  }
};

const mockHistoryPush = jest.fn();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useRouteMatch: () => ({ url: '/baseUrl' }),
  useHistory: () => ({
    push: mockHistoryPush,
  }),
}));

jest.spyOn(defaultServices.settings, 'deletionDialogsEnabledFor')
  .mockReturnValueOnce(true) // show deletion warning
  .mockReturnValue(false); // don't ask again

jest.spyOn(defaultServices.api, 'addCustomProbeTemplate').mockReturnValue(of(true));
jest.spyOn(defaultServices.api, 'deleteCustomProbeTemplate').mockReturnValue(of(true));
jest.spyOn(defaultServices.api, 'insertProbes').mockReturnValue(of(true));
jest.spyOn(defaultServices.api, 'removeProbes').mockReturnValue(of(true));

jest.spyOn(defaultServices.api, 'getProbeTemplates').mockReturnValue(of([mockCustomEventTemplate]));

jest.spyOn(defaultServices.target, 'target').mockReturnValue(of(mockTarget));
jest.spyOn(defaultServices.target, 'authFailure').mockReturnValue(of());

jest
  .spyOn(defaultServices.notificationChannel, 'messages')
  .mockReturnValueOnce(of()) // renders correctly
  .mockReturnValueOnce(of())
  
  .mockReturnValueOnce(of(mockCreateTemplateNotification)) // adds a template after receiving a notification
  .mockReturnValueOnce(of())
  .mockReturnValueOnce(of())

  .mockReturnValueOnce(of(mockDeleteTemplateNotification)) // removes a template after receiving a notification
  .mockReturnValue(of()); // all other tests

  describe('<EventTemplates />', () => {
    it('renders correctly', async () => {
      let tree;
      await act(async () => {
        tree = renderer.create(
          <ServiceContext.Provider value={defaultServices}>
            <AgentProbeTemplates />
          </ServiceContext.Provider>
        );
      });
      expect(tree.toJSON()).toMatchSnapshot();
    });

    it('adds a recording after receiving a notification', () => {
      render(
        <ServiceContext.Provider value={defaultServices}>
          <AgentProbeTemplates />
        </ServiceContext.Provider>
      );

      expect(screen.getByText('someEventTemplate')).toBeInTheDocument();
      expect(screen.getByText('anotherEventTemplate')).toBeInTheDocument();
    });

    it('removes a recording after receiving a notification', () => {
      render(
        <ServiceContext.Provider value={defaultServices}>
          <AgentProbeTemplates />
        </ServiceContext.Provider>
      );
      expect(screen.queryByText('anotherEventTemplate')).not.toBeInTheDocument();
    });

    it('displays the column header fields', () => {
      render(
        <ServiceContext.Provider value={defaultServices}>
          <AgentProbeTemplates />
        </ServiceContext.Provider>
      );
      expect(screen.getByText('name')).toBeInTheDocument();
      expect(screen.getByText('xml')).toBeInTheDocument();
    });

    it('shows a popup when uploading', () => {
      render(
        <ServiceContext.Provider value={defaultServices}>
          <AgentProbeTemplates />
        </ServiceContext.Provider>
      );
      expect(screen.queryByLabelText('Create Custom Probe Template')).not.toBeInTheDocument();

      const buttons = screen.getAllByRole('button');
      const uploadButton = buttons[0];
      userEvent.click(uploadButton);

      expect(screen.getByLabelText('Create Custom Probe Template'));

    });

    it('Tests that delete works correctly', () => {
      render(
        <ServiceContext.Provider value={defaultServices}>
          <AgentProbeTemplates />
        </ServiceContext.Provider>
      );
   
      userEvent.click(screen.getByLabelText('Actions'));

      expect(screen.getByText('Insert Probes...'));
      expect(screen.getByText('Delete'));

      const deleteAction = screen.getByText('Delete');
      userEvent.click(deleteAction);

      expect(screen.getByLabelText('Event template delete warning'));

      const deleteRequestSpy = jest.spyOn(defaultServices.api, 'deleteCustomProbeTemplate');

      expect(deleteRequestSpy).toHaveBeenCalledTimes(1);
      expect(deleteRequestSpy).toBeCalledWith('someEventTemplate');;
    });
  });
