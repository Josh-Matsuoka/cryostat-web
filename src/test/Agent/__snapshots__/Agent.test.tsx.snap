// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AgentProbeTemplates /> renders correctly 1`] = `
Array [
  <div
    className="pf-c-toolbar"
    data-ouia-component-id="OUIA-Generated-Toolbar-1"
    data-ouia-component-type="PF4/Toolbar"
    data-ouia-safe={true}
    id="event-templates-toolbar"
  >
    <div
      className="pf-c-toolbar__content"
    >
      <div
        className="pf-c-toolbar__content-section"
      >
        <div
          className="pf-c-toolbar__group pf-m-filter-group"
        >
          <div
            className="pf-c-toolbar__item"
          >
            <input
              aria-invalid={false}
              aria-label="Probe template filter"
              className="pf-c-form-control"
              data-ouia-component-id="OUIA-Generated-TextInputBase-1"
              data-ouia-component-type="PF4/TextInput"
              data-ouia-safe={true}
              disabled={false}
              id="templateFilter"
              name="templateFilter"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder="Filter..."
              readOnly={false}
              required={false}
              type="search"
            />
          </div>
        </div>
        <div
          className="pf-c-toolbar__group pf-m-icon-button-group"
        >
          <div
            className="pf-c-toolbar__item"
          >
            <button
              aria-disabled={false}
              aria-label={null}
              className="pf-c-button pf-m-primary"
              data-ouia-component-id="OUIA-Generated-Button-primary-1"
              data-ouia-component-type="PF4/Button"
              data-ouia-safe={true}
              disabled={false}
              onClick={[Function]}
              role={null}
              type="button"
            >
              Upload
            </button>
          </div>
        </div>
      </div>
      <div
        className="pf-c-toolbar__expandable-content"
        id="event-templates-toolbar-expandable-content-2"
      >
        <div
          className="pf-c-toolbar__group"
        />
      </div>
    </div>
    <div
      className="pf-c-toolbar__content pf-m-hidden"
      hidden={true}
    >
      <div
        className="pf-c-toolbar__group"
      />
    </div>
  </div>,
  <table
    aria-label="Probe Templates Table"
    className="pf-c-table pf-m-grid-md pf-m-compact"
    data-ouia-component-id="OUIA-Generated-Table-2"
    data-ouia-component-type="PF4/Table"
    data-ouia-safe={true}
    role="grid"
  >
    <thead
      className=""
    >
      <tr
        className=""
        data-ouia-component-id="OUIA-Generated-TableRow-1"
        data-ouia-component-type="PF4/TableRow"
        data-ouia-safe={true}
        hidden={false}
      >
        <th
          aria-sort="none"
          className="pf-c-table__sort"
          data-key={0}
          data-label="name"
          onMouseEnter={[Function]}
          scope="col"
        >
          <button
            className="pf-c-table__button"
            onClick={[Function]}
            type="button"
          >
            <div
              className="pf-c-table__button-content"
            >
              <span
                className="pf-c-table__text"
                onMouseEnter={[Function]}
              >
                name
              </span>
              <span
                className="pf-c-table__sort-indicator"
              >
                <svg
                  aria-hidden={true}
                  aria-labelledby={null}
                  fill="currentColor"
                  height="1em"
                  role="img"
                  style={
                    Object {
                      "verticalAlign": "-0.125em",
                    }
                  }
                  viewBox="0 0 256 512"
                  width="1em"
                >
                  <path
                    d="M214.059 377.941H168V134.059h46.059c21.382 0 32.09-25.851 16.971-40.971L144.971 7.029c-9.373-9.373-24.568-9.373-33.941 0L24.971 93.088c-15.119 15.119-4.411 40.971 16.971 40.971H88v243.882H41.941c-21.382 0-32.09 25.851-16.971 40.971l86.059 86.059c9.373 9.373 24.568 9.373 33.941 0l86.059-86.059c15.12-15.119 4.412-40.971-16.97-40.971z"
                  />
                </svg>
              </span>
            </div>
          </button>
        </th>
        <th
          aria-sort="none"
          className="pf-c-table__sort"
          data-key={1}
          data-label="xml"
          onMouseEnter={[Function]}
          scope="col"
        >
          <button
            className="pf-c-table__button"
            onClick={[Function]}
            type="button"
          >
            <div
              className="pf-c-table__button-content"
            >
              <span
                className="pf-c-table__text"
                onMouseEnter={[Function]}
              >
                xml
              </span>
              <span
                className="pf-c-table__sort-indicator"
              >
                <svg
                  aria-hidden={true}
                  aria-labelledby={null}
                  fill="currentColor"
                  height="1em"
                  role="img"
                  style={
                    Object {
                      "verticalAlign": "-0.125em",
                    }
                  }
                  viewBox="0 0 256 512"
                  width="1em"
                >
                  <path
                    d="M214.059 377.941H168V134.059h46.059c21.382 0 32.09-25.851 16.971-40.971L144.971 7.029c-9.373-9.373-24.568-9.373-33.941 0L24.971 93.088c-15.119 15.119-4.411 40.971 16.971 40.971H88v243.882H41.941c-21.382 0-32.09 25.851-16.971 40.971l86.059 86.059c9.373 9.373 24.568 9.373 33.941 0l86.059-86.059c15.12-15.119 4.412-40.971-16.97-40.971z"
                  />
                </svg>
              </span>
            </div>
          </button>
        </th>
        <td
          className=""
          data-key={2}
          data-label=""
          onMouseEnter={[Function]}
          scope={null}
        >
          
        </td>
      </tr>
    </thead>
    <tbody
      className=""
      role="rowgroup"
    >
      <tr
        className=""
        data-ouia-component-id="OUIA-Generated-TableRow-2"
        data-ouia-component-type="PF4/TableRow"
        data-ouia-safe={true}
        hidden={false}
        onClick={[Function]}
        onKeyDown={[Function]}
      >
        <td
          className=""
          data-key={0}
          data-label="name"
          onMouseEnter={[Function]}
        >
          someProbeTemplate
        </td>
        <td
          className=""
          data-key={1}
          data-label="xml"
          onMouseEnter={[Function]}
        >
          &lt;some&gt;&lt;dummy&gt;&lt;xml&gt;&lt;/xml&gt;&lt;/dummy&gt;&lt;/some&gt;
        </td>
        <td
          className="pf-c-table__action"
          data-key={2}
          data-label={null}
          onMouseEnter={[Function]}
          style={
            Object {
              "paddingRight": 0,
            }
          }
        >
          <div
            className="pf-c-dropdown pf-m-align-right"
            data-ouia-component-id="OUIA-Generated-Dropdown-1"
            data-ouia-component-type="PF4/Dropdown"
            data-ouia-safe={true}
          >
            <button
              aria-expanded={false}
              aria-haspopup={true}
              aria-label="Actions"
              className="pf-c-dropdown__toggle pf-m-plain"
              disabled={false}
              id="pf-dropdown-toggle-id-0"
              onClick={[Function]}
              onKeyDown={[Function]}
              type="button"
            >
              <svg
                aria-hidden={true}
                aria-labelledby={null}
                fill="currentColor"
                height="1em"
                role="img"
                style={
                  Object {
                    "verticalAlign": "-0.125em",
                  }
                }
                viewBox="0 0 192 512"
                width="1em"
              >
                <path
                  d="M96 184c39.8 0 72 32.2 72 72s-32.2 72-72 72-72-32.2-72-72 32.2-72 72-72zM24 80c0 39.8 32.2 72 72 72s72-32.2 72-72S135.8 8 96 8 24 40.2 24 80zm0 352c0 39.8 32.2 72 72 72s72-32.2 72-72-32.2-72-72-72-72 32.2-72 72z"
                />
              </svg>
            </button>
          </div>
        </td>
      </tr>
    </tbody>
  </table>,
]
`;
